components/ws-manager-mk2/controllers/workspace_controller_test.go

```diff
@@ -24,110 +27,373 @@ import (
+		})
+	})
+}
+
+func expectFinalizerAndMarkBackupFailed(ws *workspacev1.Workspace, pod *corev1.Pod) {
+	// Checking for the finalizer enforces our expectation that the workspace
+	// should be waiting for a backup to be taken (or fail).
+	By("checking finalizer exists for backup")
+	Consistently(func() (bool, error) {
+		if err := k8sClient.Get(ctx, types.NamespacedName{Name: pod.GetName(), Namespace: pod.GetNamespace()}, pod); err != nil {
+			return false, err
+		}
+		return controllerutil.ContainsFinalizer(pod, gitpodPodFinalizerName), nil
+	}, duration, interval).Should(BeTrue(), "missing gitpod finalizer on pod, expected one to wait for backup to succeed")
+
+	By("signalling backup completed")
+	updateObjWithRetries(ws, true, func(ws *workspacev1.Workspace) {
+		ws.Status.Conditions = wsk8s.AddUniqueCondition(ws.Status.Conditions, metav1.Condition{
+			Type:               string(workspacev1.WorkspaceConditionBackupFailure),
+			Status:             metav1.ConditionTrue,
+			Reason:             "BackupFailed",
+			LastTransitionTime: metav1.Now(),
+		})
+	})
+}
+
+func expectWorkspaceCleanup(ws *workspacev1.Workspace, pod *corev1.Pod) {
+	By("controller removing pod finalizers")
+	Eventually(func() (int, error) {
+		if err := k8sClient.Get(ctx, types.NamespacedName{Name: pod.GetName(), Namespace: pod.GetNamespace()}, pod); err != nil {
+			if errors.IsNotFound(err) {
+				// Pod got deleted, because finalizers got removed.
+				return 0, nil
+			}
+			return 0, err
+		}
+		return len(pod.ObjectMeta.Finalizers), nil
+
+	}, timeout, interval).Should(Equal(0), "pod finalizers did not go away")
+
+	By("cleaning up the workspace pod")
+	Eventually(func() error {
+		return checkNotFound(pod)
+	}, timeout, interval).Should(Succeed(), "pod did not go away")
+
+	By("cleaning up the workspace resource")
+	Eventually(func() error {
+		return checkNotFound(ws)
+	}, timeout, interval).Should(Succeed(), "workspace did not go away")
+}
+
+// checkNotFound returns nil if the object does not exist.
+// Otherwise, it returns an error.
+func checkNotFound(obj client.Object) error {
+	err := k8sClient.Get(ctx, types.NamespacedName{Name: obj.GetName(), Namespace: obj.GetNamespace()}, obj)
+	if err == nil {
+		// Object exists, return as an error.
+		return fmt.Errorf("object exists")
+	}
+	if errors.IsNotFound(err) {
+		// Object doesn't exist, this is what we want.
+		return nil
+	}
+	return err
+}
+
+func newWorkspace(name, namespace string) *workspacev1.Workspace {
+	initializer := &csapi.WorkspaceInitializer{
+		Spec: &csapi.WorkspaceInitializer_Empty{Empty: &csapi.EmptyInitializer{}},
+	}
+	initializerBytes, err := proto.Marshal(initializer)
+	Expect(err).ToNot(HaveOccurred())
+
+	return &workspacev1.Workspace{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "workspace.gitpod.io/v1",
+			Kind:       "Workspace",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: workspacev1.WorkspaceSpec{
```

Comment:
Question, non-blocking:
Should we define limits (like for CPU, network, and IO) in the WorkspaceSpec, similar to how we do now with annotations for CPU and network?

Reply:
I'll add to the mk2 milstone in a questions section for now.

Reply:
https://www.notion.so/ws-manager-mk2-questions-f256c8180e71449e95156ba3072f3290?d=8c368dbf87df4132b950afd33e5ff0ad#52f693d5c6a3454b8dc12cb78ad7948e